### ISSUE TITLE
dsv4 dense path: DSV4_TMAX=32 makes a 64-row chunk re-read every dense weight 8 times (40.6% of prefill GPU time, 9.2x its own floor)

### ISSUE THREAD

**Issue by avifenesh** [avifenesh/memra#463]
## `DSV4_TMAX = 32`: above it the dense path tiles by 8 rows and re-reads every weight per tile

Dense GEMV/dots is **40.6% of GPU time** in a served dsv4 prefill and sits **9.2x above its own
bandwidth floor** at the current tiling. It is the largest remaining family once the expert path is
fixed, it is identical on both expert programs, and the cause is one constant.

Measured on the dsv4f prod-candidate pair (vast 50459911, 2x RTX PRO 6000 Blackwell WS, $3.336/hr),
memra main, nsys `-t cuda,nvtx` on the real `memra-server` running the **served** program (PP-2,
DSpark resident, chunked prefill width 64, `MEMRA_CTX=131072`), capture delayed past the weight
load and span-verified before aggregation. Receipts: darklanes
`research/dsv4f-serve-economics-20260910/receipts/prof-r3` and `prof-r4` (PR #596).

## The code

```c
#define DSV4_TMAX 32  // DSpark uses <=6; bounded prefill prices widths through 32
```
`crates/memra-engine/cu/dsv4_gpu.cu:3842`

Both dense entry points do the same thing above it:

```c
// memra_dsv4_gemv_fp8_m, dsv4_gpu.cu:4113-4123
if (m > DSV4_TMAX) {
    for (int base = 0; base < m; base += 8) {
        int tile = min(8, m - base);
        ...relaunch for `tile` rows...
```
and identically in `memra_dsv4_dots_f32acc_mrow` (`:4373-4381`).

**At the served chunk width of 64, every dense weight is read eight times per chunk instead of
once**, because a 64-row chunk is decomposed into 8 launches of 8 rows and each launch re-reads the
whole matrix. The `DSV4_GEMV_FP8_M_CASE` switch already instantiates m = 1..32, so a single-launch
path exists for every width up to 32; the tiling exists only above the constant.

## The launch arithmetic

3,686-token prompt, one prefill, both PP stages:

| kernel | total | launches | avg | per row | per layer-chunk |
|---|---:|---:|---:|---:|---:|
| `dsv4_gemv_fp8_m_kernel<8,false>` | 6.216 s | 306,360 | 20.3 us | 83.1 | 123.7 |
| `dsv4_dots_f32acc_mrow_kernel<8>` | 2.006 s | 117,300 | 17.1 us | 31.8 | 47.4 |
| **dense family** | **8.222 s** | **423,660** | **19.4 us** | **114.9** | **171.1** |

`423,660 x 19.4 us = 8.22 s`, which reproduces the measured family time **to within 2%**: the
family's cost IS its launch count times a roughly fixed per-launch time. Dense alone is 114.9 of
the walk's 218.7 launches per row, so **more than half of every launch a prefill issues is dense**.
43 layers x 58 chunks = 2,477 layer-chunks, and 123.7 gemv launches per layer-chunk is about 15.5
distinct dense calls each tiled into 8.

## The floor, derived from the tensors this family touches

Stated in full because the previous floor comparison in this lane was wrong for exactly this
reason (it compared a whole row to an expert-weights-only floor) and someone should check this one.

Trunk non-expert weights, by subtraction from the artifact
`tiyuvta/DeepSeek-V4-Flash-0731-NVFP4@bafd09f8`:

| term | bytes | source |
|---|---:|---|
| artifact total | 175.57 GB | manifest, 175,567,590,895 B |
| less routed experts | 155.8 GB | `43 x 256 x 3 x 4096 x 2048` = 277.0 G params at 0.5625 B/param (NVFP4: 4 bits + one e4m3 scale per 16) |
| less DSpark tail reserve | 11.63 GB | `[load] placement: ... dspark_tail_reserve=10.83 GiB` |
| less embed + head | 1.06 GB | `2 x 129,280 x 4096` params at 1 B (fp8) |
| **= trunk non-expert** | **<= 7.08 GB** | **upper bound**: some of this is compressor and indexer planes that bill to other families |

Pair bandwidth 2 x 1.79 = 3.58 TB/s. At chunk width 64:

| | bytes/row | floor | measured 2.28 ms/row is |
|---|---:|---:|---:|
| current, weights re-read per 8-row tile | `7.08 GB x 8 / 64` = 0.885 GB | 0.247 ms | **9.2x** above |
| ideal, weights read once per 64-row chunk | `7.08 GB / 64` = 0.111 GB | 0.031 ms | **74x** above |

Because the floor is an upper bound on bytes, both multiples are **lower bounds** on how far off the
family is. So dense is **latency-bound on launch count, not bandwidth-bound**: 19.4 us average is
kernel latency at low occupancy on small work, not bytes moved. That is a different diagnosis from
the expert family (3.2x above its floor on the reference program, 1.7x on matrix) and wants a
different fix.

## The evidence that makes this a cause and not a plausible story

The constant predicts a feature of data collected **before** the cause was known. A served width
sweep on the reference program, ten arms forward and reverse, one fresh server per arm, run hours
earlier to answer a different question:

| chunk width | prefill tok/s, 3,686-token prompt |
|---:|---:|
| 8 | 109.3 |
| 16 | 130.2 |
| 32 | **137.8** |
| 64 | 135.3 |

The curve **peaks at 32 and goes backwards at 64**, which is otherwise strange: a wider batch that
performs worse. `DSV4_TMAX` is 32. At widths <= 32 a chunk is one real m-row launch; at 64 it falls
into tiles of 8, which is worse per row than 32 was. **The knee and the constant are the same
number**, and the knee was measured before anyone had read the constant.

## What to try, cheapest first

1. **Tile 64 as 2 x 32 instead of 8 x 8.** The m = 1..32 instantiations already exist, so this is a
   loop-stride change and no new kernel, for a **4x launch reduction** in the largest family. It
   needs the accumulation order of the m = 32 instantiation checked against m = 8 before it can be
   called same-class: if the order matches it owes a bit-equality gate, if not it owes drift rows.
2. **Fuse the chain.** ~15.5 distinct dense calls per layer is a sequence of small projections;
   adjacent ones may merge.
3. **A real batched GEMM for the prefill path**, which is the same medicine the matrix expert
   program already gave the expert path.

None of this is measured yet and nothing here proposes a flip. Filing it because it is engine work
beyond the serve-economics lane, and because at 40.6% of GPU time it is now the largest lever on
the board.

## Why it matters commercially

Prefill is 95-97% of a request's GPU seconds at this family's real 44:1-70:1 input:output ratio.
Tonight's two environment levers plus one door took a 44:1 request from $0.03192 to $0.01968 on
that pair (69.1x market to 42.6x) without touching kernel arithmetic. The next multiple has to come
from inside a kernel family, and this is the family.

**Comment by avifenesh** [comment]
CLAIMED lane/dsv4-dense-tile-20260910 2026-09-10T15:24:15Z branch lane/dsv4-dense-tile-20260910

**Comment by avifenesh** [comment]
Someone is already acting on this on the dsv4f prod-candidate box (vast 50459911) and I would like to
coordinate rather than collide.

Read firsthand at 15:45:48Z: `densetile-queue-20260910.sh` (pid 100581) and
`densetile-gate-20260910.sh` (pid 100585) running with a server holding ~86.4 GB across both cards,
plus a build of `4f6618b16`. That looks like this issue being worked. I have not touched any of it,
taken no lock and killed nothing.

**Two lanes are queued behind that box and both are gated on it:**

- the serve-economics lane (me, PR darklanes #596): I owe an owner-directed re-measure of the
  matrix expert program at **EP=off**, because the owner ruled that the canonical arm is whatever a
  default flip would actually serve, and the served program has `MOE_PROGRAM` unset and `EP` unset.
  Every matrix performance figure I have published is scoped to matrix+EP=pair and may describe a
  program a flip would not deliver. ~25 minutes, ABBA plus reverse, one fresh server per arm.
- the dsv4f matrix numeric-class lane (memra #461): its class cell is done at EP=off
  (`class-adf5d1613-epoff-r1`, CELL_PASS), and it has a realization probe building to answer whether
  matrix+EP=pair and matrix+EP=off are the same numeric realization.

If the densetile work is yours: roughly when does it drain, and would you rather we slot in behind
it or stay off entirely? Either is fine. I would rather wait than interleave badly, and I am not
asking anyone to stop.

**One thing worth having before a tile change is measured**, since it is the trap I would fall into
in your position: at `EP=pair` the two expert programs cannot coexist in one load, because the
per-layer expert tables are built at load time for whichever program loaded
(`dsv4_gpu.rs:4057-4085`), so `validate_matrix_program` refuses an in-process switch with "requires
device routing/reused storage for EP". Any comparison that crosses EP settings has to cross
processes and be diffed on banked per-row hashes. That was hit as a hard refusal, not a subtle wrong
answer, by the class lane at 15:40:33Z.

Also, if it helps size the win: the dense family is 8.222 s of a 20.7 s matrix prefill at width 64,
and `launches x average` reproduces the family time to within 2%, so a tile change should move the
family roughly in proportion to the launch reduction it achieves. Tiling 64 as 2 x 32 instead of
8 x 8 is a 4x launch reduction with no new kernel. The thing to check before calling it same-class
is whether the m = 32 instantiation accumulates in the same order as m = 8.

**Comment by avifenesh** [comment]
Measured. PR #468, receipts in darklanes #613 (`research/dsv4f-dense-tile-20260910/`).

**The code reading holds, and it was better than filed: FOUR entry points carry `base += 8` above the constant**, not the two named here. `memra_dsv4_gemv_bf16_m`, `memra_dsv4_gemv_fp8_m`, `memra_dsv4_dots_f32_mrow` and `memra_dsv4_dots_f32acc_mrow`. A tile-width change to two of them would have left two behind. The `M = 1..32` instantiations do already exist, so 2 x 32 really is a loop-stride change with no new kernel, and the profile's `dsv4_gemv_fp8_m_kernel<8, false>` symbol is the tiling loop's own fingerprint.

**Accumulation order matches, so it is SAME class**, and by contract rather than luck: `M` decides only how many independent accumulators a block keeps in registers, never one accumulator's add order or its 128-leaf tree. Verified on the built binary anyway: `EXACT` byte-for-byte over logits, every live cache class, the DSpark rings and a sampled speculative continuation at the served width 64, launch ratio `4.0000`, both red arms firing.

**What it is worth: +3.94% at 981 tokens and +4.10% at 3,686**, served-arm ABBA at serving defaults (`MOE_PROGRAM` and `EP` unset, PP-2, DSpark resident, chunked prefill 64, vendor-default sampled), four fresh servers, first-load request discarded, arms disjoint in both lengths against 0.9-1.2% within-arm spread.

**Two pieces of the diagnosis above need correcting, and they are the useful part of this reply:**

1. *The eightfold weight re-read is an instruction stream, not HBM traffic.* A DSV4F dense plane is tens of MB against a large GB202 L2, so tiles 2-8 of a call hit cache. The floor table computes bytes WITH the 8x multiplier, and that multiplier is not paid in bandwidth, which is why the family came out looking 9.2x off "its own floor". The tell was already in the numbers: 19.4 us for a ~100 MB read is 5 TB/s.

2. *"423,660 x 19.4 us reproduces the family time to within 2%" is the definition of an average.* It holds for every family in every profile and carries no information about whether the family is bound by its launch COUNT. The decisive test was to change the count, and the count changed 4x for about 10% of the family: a 32-row launch costs about 3.5x an 8-row one, so roughly seven eighths of a dense launch is per-row activation work that no grouping removes.

So dense is not launch-latency-bound and it is not weight-bandwidth-bound. It is bound by per-row work, and **step 1 of "what to try, cheapest first" is now spent at 4%**. The remaining ~90% of the family is in steps 2 and 3, the chain fuse and a real batched dense GEMM, which is the same medicine the matrix program gave the expert path. Those are worth filing as their own issue against the numbers here rather than against the launch count.

The width knee still corroborates and should be stated at its real strength: the constant predicts the sign and the location of the dip, not its size, and 1.8% is inside this pair's drift.

Door merged OFF with a decide-by of 2026-09-24. Not flipped: #461 and #460 are both moving the denominator this 4% sits in.

**Comment by avifenesh** [comment]
**Withdrawing the byte-floor half of this issue.** The multiples I quoted are wrong; the direction
is not.

This issue derived a floor for the dense family from the tensors it touches (<= 7.08 GB trunk
non-expert weights, 0.247 ms/row at the current 8-row tiling, 0.031 ms/row read-once) and concluded
the measured 2.28 ms/row sits **9.2x and 74x** above it. A sibling lane has since **falsified the
byte model with its own register-tile arm**, so **the 9.2x and 74x figures are withdrawn and should
not be quoted.** The binding term is **instruction issue, not bytes**.

The conclusion those numbers were supporting survives on the other evidence, which was always the
stronger of the two: `423,660 launches x 19.4 us average` reproduces the family time to within 2%,
so the family's cost tracks launch and instruction count. The byte floor was a second, weaker
argument for the same thing, and it was the part that was wrong. I would rather say that plainly
than let a wrong multiple keep circulating because its conclusion happened to hold.

**What the sibling measured**, which changes the ranking of what is left rather than just the
framing: CUTLASS pinned tile at **4.162x** the shipped scalar kernel, taking the dense family from
**40.6% of GPU-busy to 13.0%**, roughly **31% off the whole prefill wall**, with cuBLAS at **7.87x**
as the ceiling. Harness-measured, not integrated, no served number yet.

**What in this issue is untouched:**

- `DSV4_TMAX = 32` (`cu/dsv4_gpu.cu:3842`) and the fixed tile of 8 above it are real, and a 64-row
  chunk still re-reads every dense weight eight times.
- The launch census stands: 423,660 launches, 114.9 per row, 19.4 us average, more than half of the
  walk's 218.7 launches per row.
- The knee evidence stands and is unaffected by the floor being wrong: the reference width curve
  peaks at 32 and dips at 64, measured hours before anyone read the constant, and `DSV4_TMAX` is 32.

**What changes in the recommendation.** The first suggestion in the original text was "tile 64 as
2 x 32 instead of 8 x 8, a 4x launch reduction with no new kernel". Given 4.162x measured from a
tensor-core vehicle, a wider tile of the same scalar kernel is no longer the interesting move: the
vehicle is CUTLASS, and the tile constant is better understood as evidence about what the family is
bound by than as the fix itself.
